src/Compilers/Core/Portable/Text/SubText.cs

```diff
@@ -89,5 +94,126 @@ private TextSpan GetCompositeSpan(int start, int length)
+
+                    if (_endsWithinSplitCRLF && lineNumber == _lineCount - 1)
+                    {
+                        // Special case splitting the CRLF at the end as the UnderlyingText doesn't view the position
+                        // after between the \r and \n as on a new line whereas this subtext doesn't contain the \n
+                        // and needs to view that position as on a new line.
+                        return TextLine.FromSpanUnsafe(_subText, new TextSpan(_subText.UnderlyingSpan.End, 0));
+                    }
+
+                    var underlyingTextLine = _subText.UnderlyingText.Lines[lineNumber + _startLineNumberInUnderlyingText];
+
+                    // Consider input "a\r\nb" where ST1 contains "\a\r" and ST2 contains "\n\b", and requested lineNumber
+                    // per this table:
+                    // ----------------------------------------------------------------------------------------------------------------
+                    // | SubText | lineNumber | underlyingTextLine | _subText          | underlyingTextLine       | _subText          |
+                    // |         |            |   .Start           |   .UnderlyingSpan |   .EndIncludingLineBreak |   .UnderlyingSpan |
+                    // |         |            |                    |   .Start          |                          |   .End            |
+                    // |---------------------------------------------------------------------------------------------------------------
+                    // |   ST1   |     0      |         0          |         0         |            3             |         2         |
+                    // |   ST2   |     0      |         0          |         2         |            3             |         4         |
+                    // |   ST2   |     1      |         3          |         2         |            4             |         4         |
+                    // ----------------------------------------------------------------------------------------------------------------
+                    var startInUnderlyingText = Math.Max(underlyingTextLine.Start, _subText.UnderlyingSpan.Start);
+                    var endInUnderlyingText = Math.Min(underlyingTextLine.EndIncludingLineBreak, _subText.UnderlyingSpan.End);
+
+                    var startInSubText = startInUnderlyingText - _subText.UnderlyingSpan.Start;
+                    var length = endInUnderlyingText - startInUnderlyingText;
+
+                    var resultLine = TextLine.FromSpanUnsafe(_subText, new TextSpan(startInSubText, length));
+
+#if DEBUG
+                    // Assert resultLine only has line breaks in the appropriate locations
+                    var shouldContainLineBreak = (lineNumber != _lineCount - 1);
+                    var resultContainsLineBreak = resultLine.EndIncludingLineBreak > resultLine.End;
+
+                    Debug.Assert(shouldContainLineBreak == resultContainsLineBreak);
```

Comment:
i would make this a contract, in release mode.  and only do the ToString.All below in debug.